File: C_EXAMPLE/example_basic_Hand.cpp
/**
 * @file example_basic_Hand.cpp
 * @brief Basic example demonstrating how to control the left hand of L1 robot.
 *
 * This example shows:
 * - SDK and controller version query
 * - Hand error handling and recovery
 * - Sending predefined hand gestures
 */

#include "L1Robot.h"

#ifdef _WIN32
#define SLEEP_MS(ms) Sleep(ms)
#else
#define SLEEP_MS(ms) usleep((ms) * 1000)
#endif

 /**
  * @defgroup example_hand_control Hand Control Example
  * @{
  */

  /**
   * @brief Entry point of the basic hand control example.
   *
   * The program initializes communication with the robot system,
   * checks the left hand status, recovers from errors if needed,
   * and executes several predefined hand gestures.
   *
   * @param[in] argc Argument count (unused)
   * @param[in] argv Argument vector (unused)
   * @return int Exit code (0 on success)
   */
int main(int argc, char** argv)
{
    int sdk_version = 0;                  ///< SDK version number
    int controller_version = 0;           ///< Controller firmware version

    const ROBOT_RT* rt_ptr = FX_L1_Fbk_GetRT();

    int open_palm[24] = { 0 };            ///< Fully open hand pose
    int gesture1[24] = { 0 };             ///< First demo hand pose
    int gesture2[24] = { 0 };             ///< Second demo hand pose
    int func_ret = 0;                     ///< Function return value

    /* Initialize demo gestures */
    for (int i = 0; i < 24; i++)
    {
        gesture1[i] = 30;
        gesture2[i] = 60;
    }

    /* Query SDK version */
    sdk_version = FX_L1_System_GetSDKVersion();
    printf("SDK version is 0x%08x\n", sdk_version);

    /* Establish communication with robot system */
    if (FX_L1_System_Link(6, 6, 7, 190, FX_LOG_ALL_FLAG) < 0)
    {
        printf("Failed to link system\n");
        goto WAIT_EXIT;
    }

    /* Query controller firmware version */
    controller_version = FX_L1_System_GetControllerVersion();
    printf("Controller version is 0x%08x\n", controller_version);

    /**
     * @brief Check left hand error state
     */
    if (rt_ptr->m_HANDS[0].m_HAND_OUT.m_HAND_FBK_State == FX_HAND_STATE_ERROR)
    {
        printf("Left hand has error, press any key to reset error and enable hand\n");
        getchar();

        /* Reset left hand error */
        if (FX_L1_Runtime_SetHandAction(1, FX_HAND_LEFT, FX_HAND_ACTION_RESET) != FUNC_RET_SUCCESS)
        {
            printf("Failed to reset left hand error\n");
            goto WAIT_EXIT;
        }

        SLEEP_MS(100);

        /* Enable left hand */
        if (FX_L1_Runtime_SetHandAction(1, FX_HAND_LEFT, FX_HAND_ACTION_ENABLE) != FUNC_RET_SUCCESS)
        {
            printf("Failed to enable left hand\n");
            goto WAIT_EXIT;
        }

        SLEEP_MS(100);

        /* Re-check hand state */
        if (rt_ptr->m_HANDS[0].m_HAND_OUT.m_HAND_FBK_State == FX_HAND_STATE_ERROR)
        {
            printf("Left hand is still in error state!");
            goto WAIT_EXIT;
        }
    }

    printf("Left hand is ready to work, press any key to make a gesture\n");
    getchar();

    /* Send first gesture */
    if (FX_L1_Runtime_SetHandPos(1, FX_HAND_LEFT, gesture1) != FUNC_RET_SUCCESS)
    {
        printf("Failed to make a gesture\n");
        goto WAIT_EXIT;
    }

    printf("Press any key to make another gesture\n");
    getchar();

    /* Send second gesture */
    if (FX_L1_Runtime_SetHandPos(1, FX_HAND_LEFT, gesture2) != FUNC_RET_SUCCESS)
    {
        printf("Failed to make a gesture\n");
        goto WAIT_EXIT;
    }

    printf("Press any key to make an open palm\n");
    getchar();

    /* Open palm gesture */
    if (FX_L1_Runtime_SetHandPos(1, FX_HAND_LEFT, open_palm) != FUNC_RET_SUCCESS)
    {
        printf("Failed to make an open palm\n");
        goto WAIT_EXIT;
    }

WAIT_EXIT:
    printf("Press any key to exit\n");
    getchar();
    return 0;
}

/**
 * @}
 */

File: PYTHON_EXAMPLE/example_basic_Hand.py
"""
@file example_basic_Hand.py
@brief Basic example demonstrating how to control the left hand of L1 robot.

Workflow overview:
    1. SDK and controller version query
    2. Hand error handling and recovery
    3. Sending predefined hand gestures
"""

import sys
import time
from pathlib import Path

root_dir = Path(__file__).parent.parent
if str(root_dir) not in sys.path:
    sys.path.insert(0, str(root_dir))
from PYTHON_SDK.GentoRobot import GentoRobot, FXLogMask, FXHandType, FXHandState,FXHandAction,state_map,error_dict

def main():

    hand_type=FXHandType.FX_HAND_LEFT
    log_mask = FXLogMask.FX_LOG_INFO_FLAG
    gesture1=[30]*24
    gesture2=[60]*24
    open_palm=[0]*24

    robot = GentoRobot()

    print(f"\n### 1/5 .Connecting...")
    ret = robot.link(6, 6, 7, 190, log_level=log_mask)
    if not robot._connected:
        print(f"Link failed. Error msg: {robot._get_operate_error_msg(ret)}")
        return

    print(f"\n### 2/5. Getting versions...")
    print(f"Controller version:{robot.get_controller_version()}")
    print(f"Sdk version:{robot.get_sdk_version()}")

    print(f"\n### 3/5. Check left hand error state...")
    rt_dict = robot.get_rt_dict()
    hand_state=rt_dict["hands"][0]["fb"]["state"]
    if hand_state==FXHandState.FX_HAND_STATE_ERROR:
        ret = robot.runtime_set_hand_action(hand_type, FXHandAction.FX_HAND_ACTION_RESET)
        if ret != 0:
            print(f"Set hand reset failed. Error msg: {robot._get_operate_error_msg(ret)}")
            return
        time.sleep(0.1)

        ret = robot.runtime_set_hand_action(hand_type, FXHandAction.FX_HAND_ACTION_ENABLE)
        if ret != 0:
            print(f"Set hand enable failed. Error msg: {robot._get_operate_error_msg(ret)}")
            return
        time.sleep(0.1)

        rt_dict = robot.get_rt_dict()
        hand_state = rt_dict["hands"][0]["fb"]["state"]
        if hand_state == FXHandState.FX_HAND_STATE_ERROR:
            print("Left hand still in error!")
            return

    print(f"\n### 4/5. Hand action...")
    print("Left hand is ready to work.")
    input("Press Enter key to make a gesture1")
    ret = robot.runtime_set_hand_pos(hand_type, gesture1)
    if ret != 0:
        print(f"Set hand gesture1 failed. Error msg: {robot._get_operate_error_msg(ret)}")
        return

    input("Press Enter key to make a gesture2")
    ret = robot.runtime_set_hand_pos(hand_type, gesture2)
    if ret != 0:
        print(f"Set hand gesture2 failed. Error msg: {robot._get_operate_error_msg(ret)}")
        return


    input("Press Enter key to make an open palm")
    ret = robot.runtime_set_hand_pos(hand_type, open_palm)
    if ret != 0:
        print(f"Set hand open palm failed. Error msg: {robot._get_operate_error_msg(ret)}")
        return

    print(f"\n### 5/5. Task finished.")

if __name__ == "__main__":
    main()
